### ISSUE TITLE
Mapping String to org.joda.time.LocalDate issues compiler error

### ISSUE THREAD

**Issue by MichaelRisthausAtos** [mapstruct/mapstruct#1370]
Hi,

lets say I have a source class containing a String for a credit card expiration date (in the form of MM/yyyy) and a target class containing a joda LocalDate for the same.

So I write a mapper with a mapping:

`@Mappings({
             @Mapping(source = "creditCardExpirationDate", target = "creditCardExpirationDate", dateFormat = "MM/yyyy")
     })
     public abstract Target mapSourceToTargetWithDateTime(Source source);`

This will fail during compilation:

`Given date format "MM/yyyy" is invalid. Message: "org.joda.time.format.DateTimeFormat".`

The fancy thing is: in another project with exactly the same joda and mapstruct versions and exactly the same source (String) and target (LocalDate), Mapstruct generates the mapper just fine:

`if ( source.getCreditCardExpirationDate() != null ) {
                target.setCreditCardExpirationDate( DateTimeFormat.forPattern( "MM/yyyy" ).parseLocalDate( source.getCreditCardExpirationDate() ) );
            }`

What am I missing here? :(

The used versions are 1.2.0.Final for Mapstruct and 2.9.9 for joda time.

I created a little repository with wich you should be able to reproduce the problem:
[https://github.com/lazydays79/mapstructissue](https://github.com/lazydays79/mapstructissue)

Best regards,
Michael

**Comment by MichaelRisthausAtos** [comment]
After having a deeper look into the code of mapstruct itself, I found out, that this error will occur if mapstruct doesn't find joda time (or esp. the DateTimeFormat class) in its classpath.

If I integrate mapstruct using the "old" way by attaching it to the generate-sources phase, everything works fine.

You can find it in the [generate-sources](https://github.com/lazydays79/mapstructissue/tree/generate-sources) branch in the repository mentioned above.

You can find the difference [here](https://github.com/lazydays79/mapstructissue/compare/generate-sources).

So, the issue is: With the integration of mapstruct in the pom as given in the example in the documentation, mapstruct can't find joda in its classpath (or maybe even other dependencies?).

Regards,
Michael

**Comment by MichaelRisthausAtos** [comment]
Adding joda to the annotationProcessorPaths seems to solve the problem:

`<annotationProcessorPaths>
                        <path>
                            <groupId>joda-time</groupId>
                            <artifactId>joda-time</artifactId>
                            <version>${joda-time.version}</version>
                        </path>
                        <path>
                            <groupId>org.mapstruct</groupId>
                            <artifactId>mapstruct-processor</artifactId>
                            <version>${mapstruct.version}</version>
                        </path>
                    </annotationProcessorPaths>`

Regards,
Michael

**Comment by sjaakd** [comment]
@MichaelRisthausAtos . We don't  want to have a fixed dependency on Joda. Glad to hear you found the solution. Can I proceed and close this issue? Perhaps I can still checkout the examples / documentation to see if this is mentioned.

**Comment by MichaelRisthausAtos** [comment]
Yes, from my point of view, everything is working again just fine as long as we add joda to the annotationProcessorPath. 

But I guess that it has to be at least mentioned in the MapStruct documentation.

**Comment by sjaakd** [comment]
@MichaelRisthausAtos .. I prepared an [example](https://github.com/mapstruct/mapstruct-examples/pull/40). For some reason,  I don't need to add Soda to the `<annotationProcessorPaths>`

**Comment by ChristianBandowskiAtos** [comment]
@sjaakd The issue occurs as soon as you add a mapping from a `String` to a `DateTime` with `dateFormat` in the `@Mapping`. Then you need to add `joda-time` also to the `annotationProcessorPaths`.

**Comment by chris922** [comment]
@sjaakd The issue occurs as soon as you add a mapping from a `String` to a `DateTime` with `dateFormat` in the `@Mapping`. Then you need to add `joda-time` also to the `annotationProcessorPaths`.

**Comment by sjaakd** [comment]
Ah.. I'll prepare the example later on and see if I can reproduce this.. Its strange, because the generated code als requires Joda in the current one.

**Comment by filiphr** [comment]
I think that the actual issue is that when MapStruct runs on the javac `processorpath` probably we can't be using classes from the classpath in our code. And that is what actually occurs. The compiler says yes joda is there i.e. we can use it in the generated code, but probably not load those classes. I am not that well versed in the details of the javac compiler

**Comment by sjaakd** [comment]
I checked it.. Its a problem allright. We should not try to instantiate joda classes as part of the annotation processing. And yet we do so to validate the pattern. That's inherently wrong.. @chris922 you are right. @MichaelRisthausAtos : I'll prepare a patch.
